user/telnet/telnet.1
.\" Copyright (c) 1983, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)telnet.1	8.5 (Berkeley) 3/1/94
.\"	$Id$
.\"
.Dd March 1, 1994
.Dt TELNET 1
.Os BSD 4.2
.Sh NAME
.Nm telnet
.Nd user interface to the 
.Tn TELNET
protocol
.Sh SYNOPSIS
.Nm
.Op Fl 8EFKLacdfrx
.Op Fl S Ar tos
.Op Fl X Ar authtype
.Op Fl e Ar escapechar
.Op Fl k Ar realm
.Op Fl l Ar user
.Op Fl n Ar tracefile
.Oo
.Ar host
.Op Ar port
.Oc
.Sh DESCRIPTION
The
.Nm
command
is used to communicate with another host using the 
.Tn TELNET
protocol.
If
.Nm
is invoked without the
.Ar host
argument, it enters command mode,
indicated by its prompt
.Pq Nm telnet\&> .
In this mode, it accepts and executes the commands listed below.
If it is invoked with arguments, it performs an
.Ic open
command with those arguments.
.Pp
Options:
.Bl -tag -width indent
.It Fl 8
Specifies an 8-bit data path.  This causes an attempt to
negotiate the
.Dv TELNET BINARY
option on both input and output.
.It Fl E
Stops any character from being recognized as an escape character.
.It Fl F
If Kerberos V5 authentication is being used, the
.Fl F
option allows the local credentials to be forwarded
to the remote system, including any credentials that
have already been forwarded into the local environment.
.It Fl K
Specifies no automatic login to the remote system.
.It Fl L
Specifies an 8-bit data path on output.  This causes the
.Dv BINARY
option to be negotiated on output.
.It Fl N
Prevents IP address to name lookup when destination host is given
as an IP address.
.It Fl S Ar tos
Sets the IP type-of-service (TOS) option for the telnet
connection to the value
.Ar tos,
which can be a numeric TOS value
or, on systems that support it, a symbolic
TOS name found in the
.Pa /etc/iptos
file.
.It Fl X Ar atype 
Disables the
.Ar atype
type of authentication.
.It Fl a
Attempt automatic login.
Currently, this sends the user name via the
.Ev USER
variable
of the
.Ev ENVIRON
option if supported by the remote system.
The name used is that of the current user as returned by
.Xr getlogin 2
if it agrees with the current user ID,
otherwise it is the name associated with the user ID.
.It Fl c
Disables the reading of the user's
.Pa \&.telnetrc
file.  (See the
.Ic toggle skiprc
command on this man page.)
.It Fl d
Sets the initial value of the
.Ic debug
toggle to
.Dv TRUE .
.It Fl e Ar escapechar 
Sets the initial
.Nm
escape character to
.Ar escapechar .
If
.Ar escapechar
is omitted, then
there will be no escape character.
.It Fl f
If Kerberos V5 authentication is being used, the
.Fl f
option allows the local credentials to be forwarded to the remote system.
.It Fl k Ar realm
If Kerberos authentication is being used, the
.Fl k
option requests that
.Nm
obtain tickets for the remote host in
realm
.Ar realm
instead of the remote host's realm, as determined by
.Xr krb_realmofhost 3 .
.It Fl l Ar user 
When connecting to the remote system, if the remote system
understands the
.Ev ENVIRON
option, then
.Ar user
will be sent to the remote system as the value for the variable
.Ev USER .
This option implies the
.Fl a
option.
This option may also be used with the
.Ic open
command.
.It Fl n Ar tracefile 
Opens
.Ar tracefile
for recording trace information.
See the
.Ic set tracefile
command below.
.It Fl r
Specifies a user interface similar to
.Xr rlogin 1 .
In this
mode, the escape character is set to the tilde (~) character,
unless modified by the
.Fl e
option.
.It Fl x
Turns on encryption of the data stream if possible.  This
option is not available outside of the United States and
Canada.
.It Ar host
Indicates the official name, an alias, or the Internet address
of a remote host.
.It Ar port
Indicates a port number (address of an application).  If a number is
not specified, the default
.Nm
port is used.
.El
.Pp
When in rlogin mode, a line of the form ~.  disconnects from the
remote host; ~ is the
.Nm
escape character.
Similarly, the line ~^Z suspends the
.Nm
session.
The line ~^] escapes to the normal
.Nm
escape prompt.
.Pp
Once a connection has been opened,
.Nm
will attempt to enable the
.Dv TELNET LINEMODE
option.
If this fails, then
.Nm
will revert to one of two input modes:
either \*(Lqcharacter at a time\*(Rq
or \*(Lqold line by line\*(Rq
depending on what the remote system supports.
.Pp
When 
.Dv LINEMODE
is enabled, character processing is done on the
local system, under the control of the remote system.  When input
editing or character echoing is to be disabled, the remote system
will relay that information.  The remote system will also relay
changes to any special characters that happen on the remote
system, so that they can take effect on the local system.
.Pp
In \*(Lqcharacter at a time\*(Rq mode, most
text typed is immediately sent to the remote host for processing.
.Pp
In \*(Lqold line by line\*(Rq mode, all text is echoed locally,
and (normally) only completed lines are sent to the remote host.
The \*(Lqlocal echo character\*(Rq (initially \*(Lq^E\*(Rq) may be used
to turn off and on the local echo
(this would mostly be used to enter passwords
without the password being echoed).
.Pp
If the 
.Dv LINEMODE
option is enabled, or if the
.Ic localchars
toggle is
.Dv TRUE
(the default for \*(Lqold line by line\*(Rq; see below),
the user's
.Ic quit  ,
.Ic intr ,
and
.Ic flush
characters are trapped locally, and sent as
.Tn TELNET
protocol sequences to the remote side.
If 
.Dv LINEMODE
has ever been enabled, then the user's
.Ic susp
and
.Ic eof
are also sent as
.Tn TELNET
protocol sequences,
and
.Ic quit
is sent as a 
.Dv TELNET ABORT
instead of 
.Dv BREAK .
There are options (see
.Ic toggle
.Ic autoflush
and
.Ic toggle
.Ic autosynch
below)
which cause this action to flush subsequent output to the terminal
(until the remote host acknowledges the
.Tn TELNET
sequence) and flush previous terminal input
(in the case of
.Ic quit
and
.Ic intr  ) .
.Pp
While connected to a remote host,
.Nm
command mode may be entered by typing the
.Nm
\*(Lqescape character\*(Rq (initially \*(Lq^]\*(Rq).
When in command mode, the normal terminal editing conventions are available.
.Pp
The following
.Nm
commands are available.
Only enough of each command to uniquely identify it need be typed
(this is also true for arguments to the
.Ic mode  ,
.Ic set ,
.Ic toggle  ,
.Ic unset ,
.Ic slc  ,
.Ic environ ,
and
.Ic display
commands).
.Pp
.Bl -tag -width "mode type"
.It Ic auth Ar argument ... 
The auth command manipulates the information sent through the
.Dv TELNET AUTHENTICATE
option.  Valid arguments for the
.Ic auth
command are:
.Bl -tag -width "disable type"
.It Ic disable Ar type
Disables the specified type of authentication.  To
obtain a list of available types, use the
.Ic auth disable \&?
command.
.It Ic enable Ar type
Enables the specified type of authentication.  To
obtain a list of available types, use the
.Ic auth enable \&?
command.
.It Ic status
Lists the current status of the various types of
authentication.
.El
.It Ic close
Close a
.Tn TELNET
session and return to command mode.
.It Ic display Ar argument ... 
Displays all, or some, of the
.Ic set
and
.Ic toggle
values (see below).
.It Ic encrypt Ar argument ...
The encrypt command manipulates the information sent through the
.Dv TELNET ENCRYPT
option.
.Pp
Note:  Because of export controls, the
.Dv TELNET ENCRYPT
option is not supported outside of the United States and Canada.
.Pp
Valid arguments for the
.Ic encrypt
command are:
.Bl -tag -width Ar
.It Ic disable Ar type Ic [input|output]
Disables the specified type of encryption.  If you
omit the input and output, both input and output
are disabled.  To obtain a list of available
types, use the
.Ic encrypt disable \&?
command.
.It Ic enable Ar type Ic [input|output]
Enables the specified type of encryption.  If you
omit input and output, both input and output are
enabled.  To obtain a list of available types, use the
.Ic encrypt enable \&?
command.
.It Ic input
This is the same as the
.Ic encrypt start input
command.
.It Ic -input
This is the same as the
.Ic encrypt stop input
command.
.It Ic output
This is the same as the
.Ic encrypt start output
command.
.It Ic -output
This is the same as the
.Ic encrypt stop output
command.
.It Ic start Ic [input|output]
Attempts to start encryption.  If you omit
.Ic input
and
.Ic output,
both input and output are enabled.  To
obtain a list of available types, use the
.Ic encrypt enable \&?
command.
.It Ic status
Lists the current status of encryption.
.It Ic stop Ic [input|output]
Stops encryption.  If you omit input and output,
encryption is on both input and output.
.It Ic type Ar type
Sets the default type of encryption to be used
with later
.Ic encrypt start
or
.Ic encrypt stop
commands.
.El
.It Ic environ Ar arguments... 
The
.Ic environ
command is used to manipulate the
variables that my be sent through the
.Dv TELNET ENVIRON
option.
The initial set of variables is taken from the users
environment, with only the
.Ev DISPLAY
and
.Ev PRINTER
variables being exported by default.
The
.Ev USER
variable is also exported if the
.Fl a
or
.Fl l
options are used.
.Pp
Valid arguments for the
.Ic environ
command are:
.Bl -tag -width Fl
.It Ic define Ar variable value 
Define the variable
.Ar variable
to have a value of
.Ar value.
Any variables defined by this command are automatically exported.
The
.Ar value
may be enclosed in single or double quotes so
that tabs and spaces may be included.
.It Ic undefine Ar variable 
Remove
.Ar variable
from the list of environment variables.
.It Ic export Ar variable 
Mark the variable
.Ar variable
to be exported to the remote side.
.It Ic unexport Ar variable 
Mark the variable
.Ar variable
to not be exported unless
explicitly asked for by the remote side.
.It Ic list
List the current set of environment variables.
Those marked with a
.Cm *
will be sent automatically,
other variables will only be sent if explicitly requested.
.It Ic \&?
Prints out help information for the
.Ic environ
command.
.El
.It Ic logout
Sends the
.Dv TELNET LOGOUT
option to the remote side.
This command is similar to a
.Ic close
command; however, if the remote side does not support the
.Dv LOGOUT
option, nothing happens.
If, however, the remote side does support the
.Dv LOGOUT
option, this command should cause the remote side to close the
.Tn TELNET
connection.
If the remote side also supports the concept of
suspending a user's session for later reattachment,
the logout argument indicates that you
should terminate the session immediately.
.It Ic mode Ar type 
.Ar Type
is one of several options, depending on the state of the
.Tn TELNET
session.
The remote host is asked for permission to go into the requested mode.
If the remote host is capable of entering that mode, the requested
mode will be entered.
.Bl -tag -width Ar
.It Ic character
Disable the
.Dv TELNET LINEMODE
option, or, if the remote side does not understand the
.Dv LINEMODE
option, then enter \*(Lqcharacter at a time\*(Rq mode.
.It Ic line
Enable the
.Dv TELNET LINEMODE
option, or, if the remote side does not understand the
.Dv LINEMODE
option, then attempt to enter \*(Lqold-line-by-line\*(Rq mode.
.It Ic isig Pq Ic \-isig 
Attempt to enable (disable) the 
.Dv TRAPSIG
mode of the 
.Dv LINEMODE
option.
This requires that the 
.Dv LINEMODE
option be enabled.
.It Ic edit Pq Ic \-edit 
Attempt to enable (disable) the 
.Dv EDIT
mode of the 
.Dv LINEMODE
option.
This requires that the 
.Dv LINEMODE
option be enabled.
.It Ic softtabs Pq Ic \-softtabs 
Attempt to enable (disable) the 
.Dv SOFT_TAB
mode of the 
.Dv LINEMODE
option.
This requires that the 
.Dv LINEMODE
option be enabled.
.It Ic litecho Pq Ic \-litecho 
Attempt to enable (disable) the 
.Dv LIT_ECHO
mode of the 
.Dv LINEMODE
option.
This requires that the 
.Dv LINEMODE
option be enabled.
.It Ic \&?
Prints out help information for the
.Ic mode
command.
.El
.It Xo
.Ic open Ar host
.Op Fl l Ar user
.Oo Op Fl
.Ar port Oc
.Xc
Open a connection to the named host.
If no port number
is specified,
.Nm
will attempt to contact a
.Tn TELNET
server at the default port.
The host specification may be either a host name (see
.Xr hosts  5  )
or an Internet address specified in the \*(Lqdot notation\*(Rq (see
.Xr inet 3 ) .
The
.Fl l
option may be used to specify the user name
to be passed to the remote system via the
.Ev ENVIRON
option.
When connecting to a non-standard port,
.Nm
omits any automatic initiation of
.Tn TELNET
options.  When the port number is preceded by a minus sign,
the initial option negotiation is done.
After establishing a connection, the file
.Pa \&.telnetrc
in the
users home directory is opened.  Lines beginning with a # are
comment lines.  Blank lines are ignored.  Lines that begin
without white space are the start of a machine entry.  The
first thing on the line is the name of the machine that is
being connected to.  The rest of the line, and successive
lines that begin with white space are assumed to be
.Nm
commands and are processed as if they had been typed
in manually to the
.Nm
command prompt.
.It Ic quit
Close any open
.Tn TELNET
session and exit
.Nm telnet  .
An end of file (in command mode) will also close a session and exit.
.It Ic send Ar arguments 
Sends one or more special character sequences to the remote host.
The following are the arguments which may be specified
(more than one argument may be specified at a time):
.Pp
.Bl -tag -width escape
.It Ic abort
Sends the
.Dv TELNET ABORT
(Abort
processes)
sequence.
.It Ic ao
Sends the
.Dv TELNET AO
(Abort Output) sequence, which should cause the remote system to flush
all output
.Em from
the remote system
.Em to
the user's terminal.
.It Ic ayt
Sends the
.Dv TELNET AYT
(Are You There)
sequence, to which the remote system may or may not choose to respond.
.It Ic brk
Sends the
.Dv TELNET BRK
(Break) sequence, which may have significance to the remote
system.
.It Ic ec
Sends the
.Dv TELNET EC
(Erase Character)
sequence, which should cause the remote system to erase the last character
entered.
.It Ic el
Sends the
.Dv TELNET EL
(Erase Line)
sequence, which should cause the remote system to erase the line currently
being entered.
.It Ic eof
Sends the
.Dv TELNET EOF
(End Of File)
sequence.
.It Ic eor
Sends the
.Dv TELNET EOR
(End of Record)
sequence.
.It Ic escape
Sends the current
.Nm
escape character (initially \*(Lq^\*(Rq).
.It Ic ga
Sends the
.Dv TELNET GA
(Go Ahead)
sequence, which likely has no significance to the remote system.
.It Ic getstatus
If the remote side supports the
.Dv TELNET STATUS
command,
.Ic getstatus
will send the subnegotiation to request that the server send
its current option status.
.It Ic ip
Sends the
.Dv TELNET IP
(Interrupt Process) sequence, which should cause the remote
system to abort the currently running process.
.It Ic nop
Sends the
.Dv TELNET NOP
(No OPeration)
sequence.
.It Ic susp
Sends the
.Dv TELNET SUSP
(SUSPend process)
sequence.
.It Ic synch
Sends the
.Dv TELNET SYNCH
sequence.
This sequence causes the remote system to discard all previously typed
(but not yet read) input.
This sequence is sent as
.Tn TCP
urgent
data (and may not work if the remote system is a
.Bx 4.2
system -- if
it doesn't work, a lower case \*(Lqr\*(Rq may be echoed on the terminal).
.It Ic do Ar cmd
.It Ic dont Ar cmd
.It Ic will Ar cmd
.It Ic wont Ar cmd
Sends the
.Dv TELNET DO
.Ar cmd
sequence.
.Ar Cmd
can be either a decimal number between 0 and 255,
or a symbolic name for a specific
.Dv TELNET
command.
.Ar Cmd
can also be either
.Ic help
or
.Ic \&?
to print out help information, including
a list of known symbolic names.
.It Ic \&?
Prints out help information for the
.Ic send
command.
.El
.It Ic set Ar argument value 
.It Ic unset Ar argument value 
The
.Ic set
command will set any one of a number of
.Nm
variables to a specific value or to
.Dv TRUE .
The special value
.Ic off
turns off the function associated with
the variable, this is equivalent to using the
.Ic unset
command.
The
.Ic unset
command will disable or set to
.Dv FALSE
any of the specified functions.
The values of variables may be interrogated with the
.Ic display
command.

HISTORY
+ Release v2.0-p3
! Fixed mt9v032 video driver not invalidating cache correctly


+ Release v2.0-p2
+ rc script fix: MAC address can contain big letters
+ Added kernel as .gitmodule


+ Release v2.0-p1

Webserver Boa supports cgi streaming of large files


+ Release v2.0

+ SD-Card support


user/blkfin-apps/vlc/vlc-0.8.6b/modules/control/telnet.c
/*****************************************************************************
 * telnet.c: VLM interface plugin
 *****************************************************************************
 * Copyright (C) 2000-2006 the VideoLAN team
 * $Id: telnet.c 19602 2007-04-01 01:10:23Z hartman $
 *
 * Authors: Simon Latapie <garf@videolan.org>
 *          Laurent Aimar <fenrir@videolan.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

/*****************************************************************************
 * Preamble
 *****************************************************************************/
#include <stdlib.h>                                      /* malloc(), free() */

#include <vlc/vlc.h>
#include <vlc/intf.h>

#include <vlc/input.h>

#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>

#ifdef HAVE_SYS_TIME_H
#    include <sys/time.h>
#endif

#ifdef HAVE_UNISTD_H
#   include <unistd.h>
#endif

#include "network.h"
#include "vlc_url.h"

#include "vlc_vlm.h"

#define READ_MODE_PWD 1
#define READ_MODE_CMD 2
#define WRITE_MODE_PWD 3 // when we write the word "Password:"
#define WRITE_MODE_CMD 4

/* telnet commands */
#define TEL_WILL    251
#define TEL_WONT    252
#define TEL_DO      253
#define TEL_DONT    254
#define TEL_IAC     255
#define TEL_ECHO    1

/*****************************************************************************
 * Module descriptor
 *****************************************************************************/
static int  Open ( vlc_object_t * );
static void Close( vlc_object_t * );

#define TELNETHOST_TEXT N_( "Host" )
#define TELNETHOST_LONGTEXT N_( "This is the host on which the " \
    "interface will listen. It defaults to all network interfaces (0.0.0.0)." \
    " If you want this interface to be available only on the local " \
    "machine, enter \"127.0.0.1\"." )
#define TELNETPORT_TEXT N_( "Port" )
#define TELNETPORT_LONGTEXT N_( "This is the TCP port on which this " \
    "interface will listen. It defaults to 4212." )
#define TELNETPORT_DEFAULT 4212
#define TELNETPWD_TEXT N_( "Password" )
#define TELNETPWD_LONGTEXT N_( "A single administration password is used " \
    "to protect this interface. The default value is \"admin\"." )
#define TELNETPWD_DEFAULT "admin"

vlc_module_begin();
    set_shortname( "Telnet" );
    set_category( CAT_INTERFACE );
    set_subcategory( SUBCAT_INTERFACE_CONTROL );
    add_string( "telnet-host", "", NULL, TELNETHOST_TEXT,
                 TELNETHOST_LONGTEXT, VLC_TRUE );
    add_integer( "telnet-port", TELNETPORT_DEFAULT, NULL, TELNETPORT_TEXT,
                 TELNETPORT_LONGTEXT, VLC_TRUE );
    add_string( "telnet-password", TELNETPWD_DEFAULT, NULL, TELNETPWD_TEXT,
                TELNETPWD_LONGTEXT, VLC_TRUE );
    set_description( _("VLM remote control interface") );
    add_category_hint( "VLM", NULL, VLC_FALSE );
    set_capability( "interface", 0 );
    set_callbacks( Open , Close );
vlc_module_end();

/*****************************************************************************
 * Local prototypes.
 *****************************************************************************/
static void Run( intf_thread_t * );

typedef struct
{
    int        i_mode; /* read or write */
    int        fd;
    char       buffer_read[1000]; // 1000 byte per command should be sufficient
    char      *buffer_write;
    char      *p_buffer_read;
    char      *p_buffer_write; // the position in the buffer
    int        i_buffer_write; // the number of byte we still have to send
    int        i_tel_cmd; // for specific telnet commands

} telnet_client_t;

static char *MessageToString( vlm_message_t *, int );
static void Write_message( telnet_client_t *, vlm_message_t *, char *, int );

struct intf_sys_t
{
   telnet_client_t **clients;
   int             i_clients;
   int             *pi_fd;
   vlm_t           *mediatheque;
};

/* 
 * getPort: Decide which port to use. There are two possibilities to
 * specify a port: integrated in the --telnet-host option with :PORT
 * or using the --telnet-port option. The --telnet-port option has
 * precedence. 
 * This code relies upon the fact the url.i_port is 0 if the :PORT 
 * option is missing from --telnet-host.
 */
static int getPort(intf_thread_t *p_intf, vlc_url_t url, int i_port)
{
    // Print error if two different ports have been specified
    if (url.i_port != 0  &&
        i_port != TELNETPORT_DEFAULT && 
        url.i_port != i_port )
    {
	    msg_Err( p_intf, "ignoring port %d and using %d", url.i_port,
                 i_port);
    }
    if (i_port != TELNETPORT_DEFAULT)
    {
	    return i_port;
    }
    if (url.i_port != 0)
    {
 	    return url.i_port;
    }
    return i_port;
}

/*****************************************************************************
 * Open: initialize dummy interface
 *****************************************************************************/
static int Open( vlc_object_t *p_this )
{
    intf_thread_t *p_intf = (intf_thread_t*) p_this;
    vlm_t *mediatheque;
    char *psz_address;
    vlc_url_t url;
    int i_telnetport;

    if( !(mediatheque = vlm_New( p_intf )) )
    {
        msg_Err( p_intf, "cannot start VLM" );
        return VLC_EGENERIC;
    }

    msg_Info( p_intf, "using the VLM interface plugin..." );

    i_telnetport = config_GetInt( p_intf, "telnet-port" );
    psz_address  = config_GetPsz( p_intf, "telnet-host" );

    vlc_UrlParse(&url, psz_address, 0);

    // There might be two ports given, resolve any potentially
    // conflict
    url.i_port = getPort(p_intf, url, i_telnetport);

    p_intf->p_sys = malloc( sizeof( intf_sys_t ) );
    if( ( p_intf->p_sys->pi_fd = net_ListenTCP( p_intf, url.psz_host, url.i_port ) )
                == NULL )
    {
        msg_Err( p_intf, "cannot listen for telnet" );
        vlc_UrlClean(&url);
        free( psz_address );
        free( p_intf->p_sys );
        return VLC_EGENERIC;
    }
    msg_Info( p_intf, 
              "telnet interface started on interface %s %d",
              url.psz_host, url.i_port );

    p_intf->p_sys->i_clients   = 0;
    p_intf->p_sys->clients     = NULL;
    p_intf->p_sys->mediatheque = mediatheque;
    p_intf->pf_run = Run;

    vlc_UrlClean(&url);
    free( psz_address );
    return VLC_SUCCESS;
}

/*****************************************************************************
 * Close:
 *****************************************************************************/
static void Close( vlc_object_t *p_this )
{
    intf_thread_t *p_intf = (intf_thread_t*)p_this;
    intf_sys_t    *p_sys  = p_intf->p_sys;
    int i;

    for( i = 0; i < p_sys->i_clients; i++ )
    {
        telnet_client_t *cl = p_sys->clients[i];

        net_Close( cl->fd );
        free( cl );
    }
    if( p_sys->clients != NULL ) free( p_sys->clients );

    net_ListenClose( p_sys->pi_fd );

    vlm_Delete( p_sys->mediatheque );

    free( p_sys );
}

/*****************************************************************************
 * Run: main loop
 *****************************************************************************/
static void Run( intf_thread_t *p_intf )
{
    intf_sys_t     *p_sys = p_intf->p_sys;
    struct timeval  timeout;
    char           *psz_password;

    psz_password = config_GetPsz( p_intf, "telnet-password" );

    while( !p_intf->b_die )
    {
        fd_set fds_read, fds_write;
        int    i_handle_max = 0;
        int    i_ret, i_len, fd, i;

        /* if a new client wants to communicate */
        fd = net_Accept( p_intf, p_sys->pi_fd, p_sys->i_clients > 0 ? 0 : -1 );
        if( fd > 0 )
        {
            telnet_client_t *cl;

            /* to be non blocking */
#if defined( WIN32 ) || defined( UNDER_CE )
            {
                unsigned long i_dummy = 1;
                ioctlsocket( fd, FIONBIO, &i_dummy );
            }
#else
            fcntl( fd, F_SETFL, O_NONBLOCK );
#endif
            cl = malloc( sizeof( telnet_client_t ));
            cl->i_tel_cmd = 0;
            cl->fd = fd;
            cl->buffer_write = NULL;
            cl->p_buffer_write = cl->buffer_write;
            Write_message( cl, NULL, "Password: \xff\xfb\x01", WRITE_MODE_PWD );

            TAB_APPEND( p_sys->i_clients, p_sys->clients, cl );
        }

        /* to do a proper select */
        FD_ZERO( &fds_read );
        FD_ZERO( &fds_write );

        for( i = 0 ; i < p_sys->i_clients ; i++ )
        {
            telnet_client_t *cl = p_sys->clients[i];

            if( cl->i_mode == WRITE_MODE_PWD || cl->i_mode == WRITE_MODE_CMD )
            {
                FD_SET( cl->fd , &fds_write );
            }
            else
            {
                FD_SET( cl->fd , &fds_read );
            }
            i_handle_max = __MAX( i_handle_max, cl->fd );
        }

        timeout.tv_sec = 0;
        timeout.tv_usec = 500*1000;

        i_ret = select( i_handle_max + 1, &fds_read, &fds_write, 0, &timeout );
        if( i_ret == -1 && errno != EINTR )
        {
            msg_Warn( p_intf, "cannot select sockets" );
            msleep( 1000 );
            continue;
        }
        else if( i_ret <= 0 )
        {
            continue;
        }

        /* check if there is something to do with the socket */
        for( i = 0 ; i < p_sys->i_clients ; i++ )
        {
            telnet_client_t *cl = p_sys->clients[i];

            if( FD_ISSET(cl->fd , &fds_write) && cl->i_buffer_write > 0 )
            {
                i_len = send( cl->fd , cl->p_buffer_write ,
                              cl->i_buffer_write , 0 );
                if( i_len > 0 )
                {
                    cl->p_buffer_write += i_len;
                    cl->i_buffer_write -= i_len;
                }
            }
            else if( FD_ISSET( cl->fd, &fds_read) )
            {
                int i_end = 0;
                int i_recv;

                while( (i_recv=recv( cl->fd, cl->p_buffer_read, 1, 0 )) > 0 &&
                       cl->p_buffer_read - cl->buffer_read < 999 )
                {
                    switch( cl->i_tel_cmd )
                    {
                    case 0:
                        switch( *(uint8_t *)cl->p_buffer_read )
                        {
                        case '\r':
                            break;
                        case '\n':
                            *cl->p_buffer_read = '\n';
                            i_end = 1;
                            break;
                        case TEL_IAC: // telnet specific command
                            cl->i_tel_cmd = 1;
                            cl->p_buffer_read++;
                            break;
                        default:
                            cl->p_buffer_read++;
                            break;
                        }
                        break;
                    case 1:
                        switch( *(uint8_t *)cl->p_buffer_read )
                        {
                        case TEL_WILL: case TEL_WONT:
                        case TEL_DO: case TEL_DONT:
                            cl->i_tel_cmd++;
                            cl->p_buffer_read++;
                            break;
                        default:
                            cl->i_tel_cmd = 0;
                            cl->p_buffer_read--;
                            break;
                        }
                        break;
                    case 2:
                        cl->i_tel_cmd = 0;
                        cl->p_buffer_read -= 2;
                        break;
                    }

                    if( i_end != 0 ) break;
                }

                if( cl->p_buffer_read - cl->buffer_read == 999 )
                {
                    Write_message( cl, NULL, "Line too long\r\n",
                                   cl->i_mode + 2 );
                }

                if( i_recv == 0  || ( i_recv == -1 &&  errno != EAGAIN && errno != 0 ) )
                {
                    net_Close( cl->fd );
                    TAB_REMOVE( p_intf->p_sys->i_clients ,
                                p_intf->p_sys->clients , cl );
                    free( cl );
                }
            }
        }

        /* and now we should bidouille the data we received / send */
        for( i = 0 ; i < p_sys->i_clients ; i++ )
        {
            telnet_client_t *cl = p_sys->clients[i];

            if( cl->i_mode >= WRITE_MODE_PWD && cl->i_buffer_write == 0 )
            {
               // we have finished to send
               cl->i_mode -= 2; // corresponding READ MODE
            }
            else if( cl->i_mode == READ_MODE_PWD &&
                     *cl->p_buffer_read == '\n' )
            {
                *cl->p_buffer_read = '\0';
                if( strcmp( psz_password, cl->buffer_read ) == 0 )
                {
                    Write_message( cl, NULL, "\xff\xfc\x01\r\nWelcome, "
                                   "Master\r\n> ", WRITE_MODE_CMD );
                }
                else
                {
                    /* wrong password */
                    Write_message( cl, NULL,
                                   "\r\nWrong password.\r\nPassword: ",
                                   WRITE_MODE_PWD );
                }
            }
            else if( cl->i_mode == READ_MODE_CMD &&
                     *cl->p_buffer_read == '\n' )
            {
                /* ok, here is a command line */
                if( !strncmp( cl->buffer_read, "logout", 6 ) ||
                    !strncmp( cl->buffer_read, "quit", 4 )  ||
                    !strncmp( cl->buffer_read, "exit", 4 ) )
                {
                    net_Close( cl->fd );
                    TAB_REMOVE( p_intf->p_sys->i_clients ,
                                p_intf->p_sys->clients , cl );
                    free( cl );
                }
                else if( !strncmp( cl->buffer_read, "shutdown", 8 ) )
                {
                    msg_Err( p_intf, "shutdown requested" );
                    p_intf->p_vlc->b_die = VLC_TRUE;
                }
                else
                {
                    vlm_message_t *message;

                    /* create a standard string */
                    *cl->p_buffer_read = '\0';

                    vlm_ExecuteCommand( p_sys->mediatheque, cl->buffer_read,
                                        &message );
                    Write_message( cl, message, NULL, WRITE_MODE_CMD );
                    vlm_MessageDelete( message );
                }
            }
        }
    }
}

static void Write_message( telnet_client_t *client, vlm_message_t *message,
                           char *string_message, int i_mode )
{
    char *psz_message;

    client->p_buffer_read = client->buffer_read;
    (client->p_buffer_read)[0] = 0; // if (cl->p_buffer_read)[0] = '\n'
    if( client->buffer_write ) free( client->buffer_write );

    /* generate the psz_message string */
    if( message )
    {
        /* ok, look for vlm_message_t */
        psz_message = MessageToString( message, 0 );
    }
    else
    {
        /* it is a basic string_message */
        psz_message = strdup( string_message );
    }

    client->buffer_write = client->p_buffer_write = psz_message;
    client->i_buffer_write = strlen( psz_message );
    client->i_mode = i_mode;
}

/* We need the level of the message to put a beautiful indentation.
 * first level is 0 */
static char *MessageToString( vlm_message_t *message, int i_level )
{
#define STRING_CR "\r\n"
#define STRING_TAIL "> "

    char *psz_message;
    int i, i_message = sizeof( STRING_TAIL );

    if( !message || !message->psz_name )
    {
        return strdup( STRING_CR STRING_TAIL );
    }
    else if( !i_level && !message->i_child && !message->psz_value  )
    {
        /* A command is successful. Don't write anything */
        return strdup( /*STRING_CR*/ STRING_TAIL );
    }

    i_message += strlen( message->psz_name ) + i_level * sizeof( "    " ) + 1;
    psz_message = malloc( i_message );
    *psz_message = 0;
    for( i = 0; i < i_level; i++ ) strcat( psz_message, "    " );
    strcat( psz_message, message->psz_name );

    if( message->psz_value )
    {
        i_message += sizeof( " : " ) + strlen( message->psz_value ) +
            sizeof( STRING_CR );
        psz_message = realloc( psz_message, i_message );
        strcat( psz_message, " : " );
        strcat( psz_message, message->psz_value );
        strcat( psz_message, STRING_CR );
    }
    else
    {
        i_message += sizeof( STRING_CR );
        psz_message = realloc( psz_message, i_message );
        strcat( psz_message, STRING_CR );
    }

    for( i = 0; i < message->i_child; i++ )
    {
        char *child_message =
            MessageToString( message->child[i], i_level + 1 );

        i_message += strlen( child_message );
        psz_message = realloc( psz_message, i_message );
        strcat( psz_message, child_message );
        free( child_message );
    }

    if( i_level == 0 ) strcat( psz_message, STRING_TAIL );

    return psz_message;
}
